airflow/jobs.py

```diff
@@ -1412,20 +1413,48 @@ def _process_dags(self, dagbag, dags, tis_out):
 
         models.DagStat.update([d.dag_id for d in dags])
 
-    def _process_executor_events(self):
+    @provide_session
+    def _process_executor_events(self, simple_dag_bag, session=None):
         """
         Respond to executor events.
-
-        :param executor: the executor that's running the task instances
-        :type executor: BaseExecutor
-        :return: None
         """
-        for key, executor_state in list(self.executor.get_event_buffer().items()):
+        # TODO: this shares quite a lot of code with _manage_executor_state
+
+        TI = models.TaskInstance
+        for key, state in list(self.executor.get_event_buffer(simple_dag_bag.dag_ids)
+                                   .items()):
             dag_id, task_id, execution_date = key
             self.log.info(
                 "Executor reports %s.%s execution_date=%s as %s",
-                dag_id, task_id, execution_date, executor_state
+                dag_id, task_id, execution_date, state
             )
+            if state == State.FAILED or state == State.SUCCESS:
+                qry = session.query(TI).filter(TI.dag_id == dag_id,
+                                               TI.task_id == task_id,
+                                               TI.execution_date == execution_date)
+                ti = qry.first()
+                if not ti:
+                    self.log.warning("TaskInstance %s went missing from the database", ti)
+                    continue
+
+                if ti.state == State.RUNNING or ti.state == State.QUEUED:
+                    msg = ("Executor reports task instance {} finished ({}) "
+                           "although the task says its {}. Was the task "
+                           "killed externally?".format(ti, state, ti.state))
+                    self.log.error(msg)
+                    try:
+                        simple_dag = simple_dag_bag.get_dag(dag_id)
+                        dagbag = models.DagBag(simple_dag.full_filepath)
+                        dag = dagbag.get_dag(dag_id)
+                        ti.task = dag.get_task(task_id)
+                        ti.handle_failure(msg)
+                    except Exception:
```

Comment:
Is there nothing better we can do here?

Reply:
I share this concern (see note above).

Reply:
Please do a suggestion. I don't like it either, but to have a task instance  manage its own failure  it needs the associated task. Simple_dag cannot and does not hold the operators (pickle errors), so it is required to get it from the normal DagBag. 

I very open to suggestions, but I have been pursuing different scenarios already

Reply:
It almost feels like the alert should be sent to the admin of airflow rather than the task owner, since the appropriate action involves a config change (or machine change) to increase capacity or timeout limit on DAG imports. Is this possible in some palatable way? I don't think we have this concept right now in Airflow.

Reply:
Yeah, hence the message in the logs. I mean that is in the end what logging is for?

Reply:
Yea, but where do these logs get piped to? Is it the airflow scheduler logs, or is it the DAG-specific logs? Monitoring dag-specific logs for "error" is way too chatty since people break their DAGs all the time. If it scheduler, should be fine.

Reply:
scheduler.

Reply:
Gotcha.